### PR TITLE
Allow temporary waypoint when creating new waypoint

### PR DESCRIPTION
**Pull request by sherr-berrt**
### TLDR: 

This PR:

1. Changes the type of the state toggling the panes
2. Allows for a placeholder waypoint when making new waypoints, in order to prevent unnecessary packets being sent out

---

### State Change:

Previously, we had several states with booleans controlling which info pane is open. We only want one of them to be `true` at a time, and it gets confusing to keep track of them.

Instead, we change it to a single state with a union type,   `infoPane: "waypoint" | "waypointEdit" | "algos" | null;` so that we don't have to keep track of switching all other booleans on and off.

Changes also made to relevant selectors and components

---

### Temporary Waypoint:

Previously, new waypoints are created by first sending a packet containing a placeholder waypoint, then editing it. Because there can only be one of either a temporary waypoint or active waypoint, this PR changes the `activeWaypoint` state to hold a Waypoint object instead of an id (number), and uses the state to hold a temporary new waypoint before sending it.

This prevents unnecessary packets from being sent

### CODE REVIEW

**Comment by sherry1iu** [comment]
## Other thoughts:

For next steps, I think we should combine the `ActiveWaypoint` and `ActiveNode` states - there can only be one of them at a time, and having only one overarching state prevents clashing. `ActiveWaypoint` is now saved as a Waypoint object, so we can just check types.

**Review by ajmcquilkin — COMMENTED** [review]
Looks solid, this is a great change! Soft blocking on the duplication of the `Waypoint` object in `deviceSlice.ts`, since this could cause issues if the radio updates the active waypoint while we have it selected. Other than that just minor thoughts! 

Also you say this PR closes #356, which is the ID of this PR :)

File: src/components/Map/MapInteractionPane.tsx [hunk 1]
```diff
@@ -4,29 +4,36 @@ import { useSelector, useDispatch } from "react-redux";
...
+} from "@heroicons/react/24/outline";
 
 const MapInteractionPane = () => {
   const dispatch = useDispatch();
 
+
   const newWaypoint = useSelector(selectAllowOnMapWaypointCreation());
-  const accordionShown = useSelector(selectShowAlgosAccordion());
+  const accordionShown =
+    useSelector(selectInfoPane()) == "algos" ? true : false;
```
[the remark this answers is not in the document]
Comment (sherry1iu) [reply src/components/Map/MapInteractionPane.tsx:22]: Good catch!

For the `===`, I changed it and the link makes sense, but for future reference - why does it matter if it's a strict equality? Is there any type that can be converted into "algos"?

File: src/features/device/deviceSlice.ts [hunk 2]
```diff
@@ -8,10 +8,9 @@ export interface IDeviceState {
   activeNode: MeshNode["data"]["num"] | null;
   availableSerialPorts: string[] | null;
   activeSerialPort: string | null;
-  activeWaypoint: Waypoint["id"] | null;
-  waypointEdit: boolean; // Controls if the waypoint edit menu, or the normal menu shows up on map
+  activeWaypoint: Waypoint | null;
```
[the remark this answers is not in the document]
Comment (sherry1iu) [reply src/features/device/deviceSlice.ts:11]: So the way that the WaypointEdit pane is set up, it needs a complete waypoint object. The WaypointEdit pane shows the info only for the active waypoint. I made `activeWaypoint` be Waypoints so that all of the waypoints that WaypointEdit potentially access are accessed in the same way.


The other approach I tried was to keep `activeWaypoint` as a `Waypoint['id']`, and also have a `tempWaypoint` state in redux that either stores a single complete temporary waypoint or `null`. Then for WaypointEdit we'd do a check for if the activeWaypoint ID was 0 to see if we needed to access the `tempWaypoint` state. 

I decided against this because I want to eventually combine `activeNode` and `activeWaypoint` - since there can only be one active whatever at a time, and we want to minimize the possibility of forgetting to update one when changing the other in the future. If both are just identified as ID, there could be a case where the IDs are the same for a radio and a waypoint.

Would love to hear thoughts though

[the remark this answers is not in the document]
Comment (ajmcquilkin) [reply src/features/device/deviceSlice.ts:11]: Couldn't we write the `selectActiveWaypoint` selector similarly to the `selectActiveNode` selector? I'm not positive I'm completely understanding the problem.

On the combining of `activeNode` and `activeWaypoint`, that's an interesting idea! I haven't really thought about that, but I'm wondering if we could set some field up where the field would both tell what pane was active as well as all the information needed to render that pane. Let's discuss that at some point!

File: src/features/device/deviceSlice.ts [hunk 3]
```diff
@@ -52,39 +50,35 @@ export const deviceSlice = createSlice({
...
-    },
+    setInfoPane: (
+      state,
+      action: PayloadAction<"waypoint" | "waypointEdit" | "algos" | null>
+    ) => {
+      state.infoPane = action.payload;
 
-    setShowAlgosAccordion: (state, action: PayloadAction<boolean>) => {
-      state.showAlgosAccordion = action.payload;
       if (action.payload) {
```
[the remark this answers is not in the document]
Comment (sherry1iu) [reply src/features/device/deviceSlice.ts:72]: Not sure I completely understand - 

I think we should only have one of activeWaypoint or activeNode as non-null at a time. 

If the `action.payload` is truthy (i.e. something waypoints related) then we set `activeNode=null` to make sure to prevent the info pane for `activeNode` from popping up. 

If it's falsy, then we set `activeWaypoint=null`, and the user can click a node or do something that would make `activeNode` non-null.

File: src/features/device/deviceSlice.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ajmcquilkin) [reply src/features/device/deviceSlice.ts:11]: Hmm okay so the core issue is the new waypoint needing to be stored before being sent? That flow sounds correct for selecting the active node, although it doesn't "pull" from the radio as much as it just recalls the last data it heard from the radio. I think the clarification I would make is that we don't rely on the device to store any waypoint or text data for us, we just record all of the messages that we hear from the device and store them in the Rust backend.

Now that I'm thinking about it, would the temporary waypoint even be something we would want/need in Redux? My understanding is that the new waypoint is created and transmitted from within the `WaypointEditMenu` component, in the event that we're creating a new waypoint could we just hold the waypoint within the local component state? Once it gets sent by the device, the waypoint would end up in the redux state since we echo all waypoint and text packets back to the client.

File: src/features/device/deviceSlice.ts [hunk 4]
```diff
@@ -52,39 +50,35 @@ export const deviceSlice = createSlice({
...
-    setWaypointEdit: (state, action: PayloadAction<boolean>) => {
-      state.waypointEdit = action.payload;
-    },
-
-    setAllowOnMapWaypointCreation: (state, action: PayloadAction<boolean>) => {
-      state.allowOnMapWaypointCreation = action.payload;
-    },
+    setInfoPane: (
+      state,
+      action: PayloadAction<"waypoint" | "waypointEdit" | "algos" | null>
```
[the remark this answers is not in the document]
Comment (sherry1iu) [reply src/features/device/deviceSlice.ts:68]: Sure I'd be down with that, can I ask why?

[the remark this answers is not in the document]
Comment (ajmcquilkin) [reply src/features/device/deviceSlice.ts:68]: I think the reason I had left this comment was because the type of the active pane can't be accessed by other files without using some funky `typeof` constructs. Looking back on this, my comment was misleading, I think the big issue I was having was just not having the union type not declared outside of the `setInfoPane` method. Nothing particularly special about enums, although I just found this article that you might find interesting ([link](https://www.bam.tech/article/should-you-use-enums-or-union-types-in-typescript))

File: src/features/device/deviceSlice.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ajmcquilkin) [reply src/features/device/deviceSlice.ts:72]: I guess my point of confusion was around this flow not including the `activeNode` code, since we close the node info pane by setting `activeNode` to `null`. Do you have thoughts on whether it would be possible/worth it to include the `activeNode` flow in this method?

File: src/features/device/deviceSlice.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (sherr-berrt) [reply src/features/device/deviceSlice.ts:11]: So the new waypoint is currently being created from the `MapView` component, then passed to redux. Then `WaypointEditMenu` finds the active waypoint from redux (similar to how the `MapSelectedNodeMenu` works). 

I'd push back against storing the info in a local component state - there are two components that depend on a complete temporary waypoint from `MapView`: the map pin (`Waypoint.tsx`) and the menu (`WaypointEditMenu`). If we cancel creating the waypoint from `WaypointEditMenu,` or add in emoji handling later, we'd also have to update `Waypoint.tsx`, rather than just having both get info from redux - I think it's messier than just having another state

File: src/features/device/deviceSlice.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (sherr-berrt) [reply src/features/device/deviceSlice.ts:72]: I think it would be nice to do it, but it might be better as its own PR, since the node pane is currently connected to `activeNode`, and we might be changing the type of `activeNode` to include waypoints.

File: src/features/device/deviceSlice.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (sherr-berrt) [reply src/features/device/deviceSlice.ts:11]: Time for me to (in barkin's words) cry over git again 

_(I know what to do, that was a joke)_

File: src/features/device/deviceSlice.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ajmcquilkin) [reply src/features/device/deviceSlice.ts:72]: Yeah that's fair, good to resolve this in that case

**Review by ajmcquilkin — APPROVED** [review]
